### PR TITLE
test: Add terraform test GitHub Action (and make tests pass)

### PR DESCRIPTION
**Pull request by WillGibson**
During a conversation with @antroy-madetech in our Slack channel, we agreed that the Terraform unit tests should behave the same wherever they are run from, whether you are logged into AWS or not etc. etc.

This pull request aims to achieve that.

It also adds a GitHub Action to run all the tests against the latest two minor release of Terraform.

### CODE REVIEW

File: extensions/tests/unit.tftest.hcl [hunk 1]
```diff
@@ -61,41 +77,39 @@ override_data {
 run "aws_ssm_parameter_unit_test" {
   command = plan
 
+  # Configuration
```
Comment (WillGibson) [inline extensions/tests/unit.tftest.hcl:80]: Ignore these, they are changes in readiness for what I was trying to do before I went down this rabbit hole.

File: data-migration/main.tf [hunk 2]
```diff
@@ -1,6 +1,6 @@
 resource "aws_iam_role" "s3_migration_role" {
   name               = local.role_name
-  assume_role_policy = data.aws_iam_policy_document.allow_assume_role.json
+  assume_role_policy = jsonencode(data.aws_iam_policy_document.allow_assume_role)
```
Comment (WillGibson) [inline data-migration/main.tf:3]: This works with the tests, `.json` does not.

File: extensions/tests/unit.tftest.hcl [hunk 3]
```diff
@@ -61,41 +77,39 @@ override_data {
...
     condition     = aws_ssm_parameter.addons.tier == "Intelligent-Tiering"
     error_message = "Intelligent-Tiering not enabled, parameters > 4096 characters will be rejected"
   }
-
   assert {
     condition     = aws_ssm_parameter.addons.type == "String"
     error_message = "Invalid config for aws_ssm_parameter type"
   }
 
+  # Value
```
Comment (WillGibson) [inline extensions/tests/unit.tftest.hcl:94]: We don't have a test for this yet 😬

File: .github/workflows/terraform-unit-tests.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+          do
+            # Lose leading ./ and select the part before the tests directory
+            module=$(echo "${file#./}" | awk -F "/tests/" '{print $1}')
+            # In case we separate the test files, only include each module once
+            if [[ ${modules} != *"$module"* ]]; then
+              echo "Found module $module"
+              modules+="\"$module\","
+            fi
+          done
+          echo "modules=[${modules%,}]" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (WillGibson) [reply .github/workflows/terraform-unit-tests.yml:35]: I would very much like to remove the CodeBuild jobs if we can get alerts to Slack from GitHub.

@simon106 is on that mission.

In the meantime, I'll update the CodeBuild job.

File: .github/workflows/terraform-unit-tests.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+name: Terraform Unit Tests
+
+on:
+  push:
```
[the remark this answers is not in the document]
Comment (WillGibson) [reply .github/workflows/terraform-unit-tests.yml:4]: Yes.

File: .github/workflows/terraform-unit-tests.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (WillGibson) [reply .github/workflows/terraform-unit-tests.yml:35]: https://github.com/uktrade/terraform-tools/pull/43

File: application-load-balancer/tests/unit.tftest.hcl [hunk 6]
```diff
@@ -1,3 +1,5 @@
+mock_provider "aws" {}
+
 mock_provider "aws" {
   alias = "sandbox"
 }
```
Comment (WillGibson) [inline application-load-balancer/tests/unit.tftest.hcl:5]: @chiaramapellimt If you question in Slack relates to the two providers above and the one below. 

I think the first one covers where we would be using the default providers and the others relate to specific accounts. E.g. https://github.com/uktrade/terraform-platform-modules/blob/main/application-load-balancer/main.tf#L110

Comment (chiaramapellimt) [reply application-load-balancer/tests/unit.tftest.hcl:5]: oooo I see

File: .github/workflows/terraform-unit-tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (antroy-madetech) [reply .github/workflows/terraform-unit-tests.yml:4]: Do we want that? Or just when pushed to a branch with an open PR?

[the remark this answers is not in the document]
Comment (antroy-madetech) [reply .github/workflows/terraform-unit-tests.yml:4]: Not convinced you can actually from looking at the docs, so this is fine.

[the remark this answers is not in the document]
Comment (WillGibson) [reply .github/workflows/terraform-unit-tests.yml:4]: I (personally) want fast feedback if I mess stuff up, not to wait until I perhaps create a pull request.

The Action is actually way quicker to run all the tests than doing it locally.

**Comment by WillGibson** [comment]
I am going to merge this now because I need to get on with the next thing. Feedback is still welcome.

**Pull request merged**